libvast/include/vast/system/index.hpp

```diff
@@ -141,26 +141,84 @@ struct query_backlog {
   std::deque<job> low;
 };
 
-struct query_state {
-  /// The UUID of the query.
-  vast::uuid id;
+/// The result type of the query evaluation handler.
+enum class query_response : bool {
+  /// Signals that query evaluation is incomplete and more results can be
+  /// requested.
+  incomplete,
+  /// Signals that the query is fully evaluated.
+  complete
+};
 
+struct query_state {
   /// The query expression.
   vast::query query;
 
-  /// Unscheduled partitions.
-  std::vector<uuid> partitions;
+  /// The query client.
+  receiver_actor<atom::done> client = {};
+
+  /// The number of partitions that need to be evaluated for this query.
+  uint32_t candidate_partitions = 0;
 
-  /// The assigned query worker.
-  query_supervisor_actor worker;
+  /// The number of partitions that have been reqested by the client.
+  uint32_t requested_partitions = 0;
+
+  /// The number of partitions that the query was sent to.
+  uint32_t scheduled_partitions = 0;
+
+  /// The number of partitions that are processed already.
+  uint32_t completed_partitions = 0;
+
+  /// A response promise to signal batch completion to the client.
+  caf::typed_response_promise<query_response> rp = {};
 
   template <class Inspector>
   friend auto inspect(Inspector& f, query_state& x) {
-    return f(caf::meta::type_name("query_state"), x.id, x.query,
-             caf::meta::omittable_if_empty(), x.partitions, x.worker);
+    return f(caf::meta::type_name("query_state"), x.query, x.client,
+             x.candidate_partitions, x.requested_partitions,
+             x.scheduled_partitions, x.completed_partitions);
   }
 };
 
+struct index_counters {
+  size_t partition_materializations = 0;
+  size_t partition_lookups = 0;
+};
+
+struct pending_queue {
+  struct pq {
+    uuid partition;
+    std::vector<uuid> queries;
+
+    friend auto operator<=>(const pq& lhs, const pq& rhs) {
+      return lhs.queries.size() <=> rhs.queries.size();
+    }
+  };
+
+  /// Inserts a new query into the queue.
+  [[nodiscard]] caf::error
+  insert(query_state&& query_state, std::vector<uuid>&& candidates);
+
+  /// Activates an inactive query.
+  [[nodiscard]] caf::error activate(const uuid& qid, uint32_t num_partitions);
+
+  [[nodiscard]] caf::error remove_query(const uuid& qid);
+
+  [[nodiscard]] size_t num_partitions() const;
+
+  [[nodiscard]] size_t num_queries() const;
+
+  [[nodiscard]] uuid create_query_id() const;
+
+  [[nodiscard]] std::optional<pq> next();
```

Comment:
Please document. It's hard to understand what this is doing without going through all of the code.